src/Interpreters/Access/InterpreterCreateRoleQuery.cpp

```diff
@@ -78,12 +93,27 @@ BlockIO InterpreterCreateRoleQuery::execute()
             new_roles.emplace_back(std::move(new_role));
         }
 
+        if (!query.storage_name.empty())
+        {
+            for (const auto & name : query.names)
+            {
+                auto id = access_control.find<Role>(name);
+
+                if (!id)
+                    continue;
+
+                auto another_storage_ptr = access_control.findStorage(*id);
+                if (another_storage_ptr != storage_ptr)
+                    throw Exception(ErrorCodes::ACCESS_ENTITY_ALREADY_EXISTS, "Role {} already exists in storage {}", name, another_storage_ptr->getStorageName());
+            }
+        }
```

Comment:
Maybe it's worth to move that logic to a new function named like `MultipleAccessStorage::findExcludingStorage()`